File: typescript/test/nodeProgressTracker.test.ts
import { describe, expect, test } from "vitest";

import { Dag, NodeProgressTracker } from "../src";

describe("NodeProgressTracker", () => {
  test("emptyDag_initializesNoEntries_andProgressIsZero", () => {
    const dag = Dag.builder<string>().build();
    const tracker = new NodeProgressTracker(dag);
    expect(tracker.progress).toBe(0);
  });

  test("constructor_initializesAllNodeIdsToZero", () => {
    const b = Dag.builder<string>();
    const idA = b.add("a");
    const idB = b.add("b");
    const idC = b.add("c");
    const dag = b.build();
    const tracker = new NodeProgressTracker(dag);
    expect(tracker.get(idA)).toBe(0);
    expect(tracker.get(idB)).toBe(0);
    expect(tracker.get(idC)).toBe(0);
  });

  test("putAndGet_roundTrip", () => {
    const b = Dag.builder<string>();
    const only = b.add("x");
    const dag = b.build();
    const tracker = new NodeProgressTracker(dag);
    tracker.put(only, 0.5);
    expect(tracker.get(only)).toBeCloseTo(0.5, 6);
  });

  test("put_rejectsValueBelowZero", () => {
    const b = Dag.builder<string>();
    const id = b.add("x");
    const dag = b.build();
    const tracker = new NodeProgressTracker(dag);
    expect(() => tracker.put(id, -0.01)).toThrowError(/Progress must be between 0 and 1\./);
  });

  test("put_rejectsValueAboveOne", () => {
    const b = Dag.builder<string>();
    const id = b.add("x");
    const dag = b.build();
    const tracker = new NodeProgressTracker(dag);
    expect(() => tracker.put(id, 1.01)).toThrow();
  });

  test("put_acceptsBoundaryValues", () => {
    const b = Dag.builder<string>();
    const lo = b.add("lo");
    const hi = b.add("hi");
    const dag = b.build();
    const tracker = new NodeProgressTracker(dag);
    tracker.put(lo, 0);
    tracker.put(hi, 1);
    expect(tracker.get(lo)).toBe(0);
    expect(tracker.get(hi)).toBe(1);
  });

  test("progress_isAverageAcrossNodes", () => {
    const b = Dag.builder<string>();
    const n0 = b.add("a");
    const n1 = b.add("b");
    const n2 = b.add("c");
    const n3 = b.add("d");
    const dag = b.build();
    const tracker = new NodeProgressTracker(dag);
    tracker.put(n0, 0);
    tracker.put(n1, 0.5);
    tracker.put(n2, 1);
    tracker.put(n3, 0.25);
    expect(tracker.progress).toBeCloseTo(0.4375, 6);
  });

  test("customMapSupplier_isPopulatedAndUsed", () => {
    const b = Dag.builder<string>();
    const first = b.add("a");
    const second = b.add("b");
    const dag = b.build();
    const backing = new Map<number, number>();
    const tracker = new NodeProgressTracker(dag, () => backing);
    expect(backing.size).toBe(2);
    expect(backing.get(first)).toBe(0);
    expect(backing.get(second)).toBe(0);
    tracker.put(first, 1);
    tracker.put(second, 0);
    expect(tracker.progress).toBeCloseTo(0.5, 6);
    expect(backing.get(first)).toBe(1);
  });
});

File: java/src/test/java/io/github/flashlock/NodeProgressTrackerTest.java
package io.github.flashlock;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.LinkedHashMap;
import java.util.Map;
import org.junit.jupiter.api.Test;

class NodeProgressTrackerTest {

  @Test
  void emptyDag_initializesNoEntries_andProgressIsZero() {
    Dag<String> dag = Dag.<String>builder().build();
    NodeProgressTracker tracker = new NodeProgressTracker(dag);
    assertEquals(0f, tracker.progress(), 1e-6f);
  }

  @Test
  void constructor_initializesAllNodeIdsToZero() {
    Dag.Builder<String> b = Dag.<String>builder();
    int idA = b.add("a");
    int idB = b.add("b");
    int idC = b.add("c");
    Dag<String> dag = b.build();
    NodeProgressTracker tracker = new NodeProgressTracker(dag);
    assertEquals(0f, tracker.get(idA), 0f);
    assertEquals(0f, tracker.get(idB), 0f);
    assertEquals(0f, tracker.get(idC), 0f);
  }

  @Test
  void putAndGet_roundTrip() {
    Dag.Builder<String> b = Dag.<String>builder();
    int only = b.add("x");
    Dag<String> dag = b.build();
    NodeProgressTracker tracker = new NodeProgressTracker(dag);
    tracker.put(only, 0.5f);
    assertEquals(0.5f, tracker.get(only), 1e-6f);
  }

  @Test
  void put_rejectsValueBelowZero() {
    Dag.Builder<String> b = Dag.<String>builder();
    int id = b.add("x");
    Dag<String> dag = b.build();
    NodeProgressTracker tracker = new NodeProgressTracker(dag);
    IllegalArgumentException ex =
        assertThrows(IllegalArgumentException.class, () -> tracker.put(id, -0.01f));
    assertTrue(ex.getMessage().startsWith("Progress must be between 0 and 1."));
  }

  @Test
  void put_rejectsValueAboveOne() {
    Dag.Builder<String> b = Dag.<String>builder();
    int id = b.add("x");
    Dag<String> dag = b.build();
    NodeProgressTracker tracker = new NodeProgressTracker(dag);
    assertThrows(IllegalArgumentException.class, () -> tracker.put(id, 1.01f));
  }

  @Test
  void put_acceptsBoundaryValues() {
    Dag.Builder<String> b = Dag.<String>builder();
    int lo = b.add("lo");
    int hi = b.add("hi");
    Dag<String> dag = b.build();
    NodeProgressTracker tracker = new NodeProgressTracker(dag);
    tracker.put(lo, 0f);
    tracker.put(hi, 1f);
    assertEquals(0f, tracker.get(lo), 0f);
    assertEquals(1f, tracker.get(hi), 0f);
  }

  @Test
  void progress_isAverageAcrossNodes() {
    Dag.Builder<String> b = Dag.<String>builder();
    int n0 = b.add("a");
    int n1 = b.add("b");
    int n2 = b.add("c");
    int n3 = b.add("d");
    Dag<String> dag = b.build();
    NodeProgressTracker tracker = new NodeProgressTracker(dag);
    tracker.put(n0, 0f);
    tracker.put(n1, 0.5f);
    tracker.put(n2, 1f);
    tracker.put(n3, 0.25f);
    assertEquals(0.4375f, tracker.progress(), 1e-6f);
  }

  @Test
  void customMapSupplier_isPopulatedAndUsed() {
    Dag.Builder<String> b = Dag.<String>builder();
    int first = b.add("a");
    int second = b.add("b");
    Dag<String> dag = b.build();
    Map<Integer, Float> backing = new LinkedHashMap<>();
    NodeProgressTracker tracker = new NodeProgressTracker(dag, () -> backing);
    assertEquals(2, backing.size());
    assertEquals(0f, backing.get(first), 0f);
    assertEquals(0f, backing.get(second), 0f);
    tracker.put(first, 1f);
    tracker.put(second, 0f);
    assertEquals(0.5f, tracker.progress(), 1e-6f);
    assertEquals(1f, backing.get(first), 0f);
  }
}
